imp_developer.py

```diff
@@ -272,56 +326,414 @@ def __base64_encode(str):
         return base64.b64encode(str.encode()).decode()
 
     @staticmethod
-    def __get_http_headers(key):
-        return {
-            "Authorization": "Basic " + HTTP.__base64_encode(key),
-            "Content-Type": "application/json",
-            "User-Agent": "imp-developer/sublime"
-        }
+    def get_http_headers(key, headers):
+        if not headers:
+            headers = {}
+        if key:
+            headers["Authorization"] = "Bearer " + key
+
+        return headers
 
     @staticmethod
-    def is_build_api_key_valid(key):
-        request, code = HTTP.get(key, PL_BUILD_API_URL_V4 + "models")
+    def is_refresh_token_valid(token):
+        if datetime.datetime.strptime(token.get(EI_EXPIRES_AT), "%Y-%m-%dT%H:%M:%S.%fZ") > datetime.now():
+            return True
+        request, code = HTTP.post({token:token.value}, PL_IMPCENTRAL_API_URL_V5 + "/auth/token")
         return code == 200
 
+
     @staticmethod
-    def do_request(key, url, method, data=None, timeout=None):
+    def do_request(key, url, method, data=None, timeout=None, headers=None):
         if data:
             data = data.encode('utf-8')
         req = urllib.request.Request(url,
-                                     headers=HTTP.__get_http_headers(key),
+                                     headers=HTTP.get_http_headers(key, headers),
                                      data=data,
                                      method=method)
 
         result, code = None, None
         try:
-            res = urllib.request.urlopen(req, timeout=timeout)
+            res = urllib.request.urlopen(req, timeout=None)
             code = res.getcode()
-            result = json.loads(res.read().decode('utf-8'))
+            pl = res.read().decode('utf-8')
+            if pl:
+                result = json.loads(pl)
         except socket.timeout:
             log_debug("Timeout error occurred for URL: " + url)
             result = code
         except urllib.error.HTTPError as err:
             code = err.code
             result = json.loads(err.read().decode('utf-8'))
+        except urllib.error.URLError as err:
+            code = 404
+            result = {"error" : err.reason}
+        except urllib.error.ContentTooShortError as err:
+            code = 404
```

Comment:
Should it be just 400?

Reply:
probably, it could be 404 or 400, but it is it easier to handle one error code instead of multiple

Reply:
So, it is trick to avoid detailed error analysis